contracts/PangolinExchange.sol

```diff
@@ -0,0 +1,166 @@
+
+    address tokenAddress = supportedAssets.getAssetAddress(_token);
+
+    uint256 amountIn = getEstimatedAVAXForERC20Token(_exactERC20AmountOut, tokenAddress);
+
+    require(msg.value >= amountIn, "Not enough funds provided");
+
+    pangolinRouter.swapAVAXForExactTokens{value : msg.value}(_exactERC20AmountOut, getPathForAVAXtoToken(tokenAddress), msg.sender, block.timestamp);
+
+    emit TokenPurchase(msg.sender, _exactERC20AmountOut, block.timestamp);
+  }
+
+
+  /**
+   * Sells selected ERC20 token for AVAX
+   * @dev _token ERC20 token's address
+   * @dev _exactERC20AmountIn amount of the ERC20 token to be sold
+   * @dev _minAvaxAmountOut minimum amount of the AVAX token to be bought
+  **/
+  function sellAsset(bytes32 _token, uint256 _exactERC20AmountIn, uint256 _minAvaxAmountOut) external override RefundRemainder {
+    require(_exactERC20AmountIn > 0, "Amount of tokens to sell has to be greater than 0");
+    address tokenAddress = supportedAssets.getAssetAddress(_token);
+
+    IERC20 token = IERC20(tokenAddress);
+    token.approve(address(pangolinRouter), _exactERC20AmountIn);
+    pangolinRouter.swapExactTokensForAVAX(_exactERC20AmountIn, _minAvaxAmountOut, getPathForTokenToAVAX(tokenAddress), msg.sender, block.timestamp);
+
+    emit TokenSell(msg.sender, _exactERC20AmountIn, block.timestamp);
+  }
+
+  /* ========== RECEIVE AVAX FUNCTION ========== */
+  receive() external payable {  }
+
+
+  /* ========== VIEW FUNCTIONS ========== */
+
+  /**
+    * Returns the current balance of the asset held by a given user
+    * @dev _asset the code of an asset
+    * @dev _user the address of queried user
+  **/
+  function getBalance(address _user, bytes32 _asset) external override view returns(uint256) {
+    IERC20 token = IERC20(supportedAssets.getAssetAddress(_asset));
+    return token.balanceOf(_user);
+  }
+
+
+  /**
+     * Returns the minimum token amount that is required to be sold to receive _exactAmountOut of AVAX.
+  **/
+  function getMinimumERC20TokenAmountForExactAVAX(uint256 _exactAmountOut, address _token) public view override returns (uint256) {
+    address[] memory path = getPathForTokenToAVAX(_token);
+
+    return pangolinRouter.getAmountsIn(_exactAmountOut, path)[0];
+  }
+
+
+  /**
+     * Returns the minimum AVAX amount that is required to buy _exactAmountOut of _token ERC20 token.
+  **/
+  function getEstimatedAVAXForERC20Token(uint256 _exactAmountOut, address _token) public view returns (uint256) {
+    address[] memory path = getPathForAVAXtoToken(_token);
+
+    return pangolinRouter.getAmountsIn(_exactAmountOut, path)[0];
+  }
+
+  /**
+   * Returns the maximum AVAX amount that will be obtained in the event of selling _amountIn of _token ERC20 token.
+  **/
+  function getEstimatedAVAXFromERC20Token(uint256 _amountIn, address _token) public view override returns (uint256) {
+    address[] memory path = getPathForTokenToAVAX(_token);
+
+    return pangolinRouter.getAmountsOut(_amountIn, path)[1];
```

Comment:
Pls check if it might throw an exception

Reply:
@kamilovsky01, could you check it?

Reply:
The answer is exactly the same as in the question above. From our end we should be safe. What about the reverves-related errors?